crates/query-engine/translation/src/translation/query/fields.rs

```diff
@@ -318,3 +323,201 @@ fn translate_nested_field(
+    expression: sql::ast::Expression,
+    column_type_representation: Option<TypeRepresentation>,
+) -> sql::ast::Expression {
+    match column_type_representation {
+        None => expression,
+        Some(type_rep) => {
+            if let Some(sql::ast::ScalarType(cast_type)) =
+                get_type_representation_cast_type(&type_rep)
+            {
+                sql::ast::Expression::Cast {
+                    expression: Box::new(expression),
+                    // make it an array of cast type
+                    r#type: sql::ast::ScalarType(format!("{cast_type}[]")),
+                }
+            } else {
+                expression
+            }
+        }
+    }
+}
+
+/// Certain type representations require that we provide a different json representation
+/// than what postgres will return.
+/// For columns of those type representation, we wrap the result in a cast.
+fn wrap_in_type_representation(
+    expression: sql::ast::Expression,
+    column_type_representation: Option<TypeRepresentation>,
+) -> sql::ast::Expression {
+    match column_type_representation {
+        None => expression,
+        Some(type_rep) => {
+            if let Some(cast_type) = get_type_representation_cast_type(&type_rep) {
+                sql::ast::Expression::Cast {
+                    expression: Box::new(expression),
+                    r#type: cast_type,
+                }
+            } else {
+                expression
+            }
+        }
+    }
+}
+
+/// If a type representation requires a cast, return the scalar type name.
+fn get_type_representation_cast_type(
+    type_representation: &TypeRepresentation,
+) -> Option<sql::ast::ScalarType> {
+    match type_representation {
+        // In these situations, we expect to cast the expression according
+        // to the type representation.
+        TypeRepresentation::Int64AsString => Some(sql::ast::ScalarType("text".to_string())),
+        TypeRepresentation::BigDecimalAsString => Some(sql::ast::ScalarType("text".to_string())),
+
+        // In these situations the type representation should be the same as
+        // the expression, so we don't cast it.
+        TypeRepresentation::Boolean => None,
+        TypeRepresentation::String => None,
+        TypeRepresentation::Float32 => None,
+        TypeRepresentation::Float64 => None,
+        TypeRepresentation::Int16 => None,
+        TypeRepresentation::Int32 => None,
+        TypeRepresentation::Int64 => None,
+        TypeRepresentation::BigDecimal => None,
+        TypeRepresentation::Timestamp => None,
+        TypeRepresentation::Timestamptz => None,
+        TypeRepresentation::Time => None,
+        TypeRepresentation::Timetz => None,
+        TypeRepresentation::Date => None,
+        TypeRepresentation::UUID => None,
+        TypeRepresentation::Geography => None,
+        TypeRepresentation::Geometry => None,
+        TypeRepresentation::Json => None,
+        TypeRepresentation::Enum(_) => None,
```

Comment:
I appreciate you listing these out rather than using a wildcard.

You can use `|` to join them, though (there's a pedantic warning for this but we haven't enabled it).

Reply:
cool. Thanks!